metagraph/src/cli/assemble.cpp

```diff
@@ -93,15 +227,62 @@ int assemble(Config *config) {
 
     logger->trace("Graph loaded in {} sec", timer.elapsed());
 
-    std::unique_ptr<graph::AnnotatedDBG> anno_graph;
     if (config->infbase_annotators.size()) {
-        anno_graph = initialize_annotated_dbg(graph, *config);
+        config->infbase = files.at(0);
+
+        assert(config->label_mask_file.size());
+        auto anno_graph = initialize_annotated_dbg(graph, *config);
+
+        logger->trace("Generating masked graphs...");
+
+        std::filesystem::remove(
+            utils::remove_suffix(config->outfbase, ".gz", ".fasta") + ".fasta.gz"
+        );
+
+        std::mutex write_mutex;
+
+        assert(!config->label_mask_file.empty());
 
-        logger->trace("Masking graph...");
+        std::ifstream fin(config->label_mask_file);
+        if (!fin.good()) {
+            throw std::iostream::failure("Failed to read label mask file");
+            exit(1);
+        }
 
-        graph = mask_graph(*anno_graph, config);
+        auto [lines, num_experiments] = parse_diff_file(fin);
+        if (!num_experiments)
+            return 0;
+
+        size_t num_threads = std::max(1u, get_num_threads());
+
+        call_masked_graphs(*anno_graph, config, lines,
+            [&](const graph::MaskedDeBruijnGraph &graph, const std::string &header) {
+                seq_io::FastaWriter writer(config->outfbase, header,
+                                           config->enumerate_out_sequences,
+                                           num_threads > 1, /* async write */
+                                           "a" /* append mode */);
+
+                if (config->unitigs || config->min_tip_size > 1) {
+                    graph.call_unitigs([&](const auto &unitig, auto&&) {
+                                           std::lock_guard<std::mutex> lock(write_mutex);
```

Comment:
Yes. async mode simply means that writes to disk happen in a separate thread so pushes to the buffer can continue to happen concurrently. The lock here is needed since pushes are not thread safe

Reply:
Ah, got it. It's a different kind of parallelism. Maybe calling `async_writer` instead of the simple `async` would help to avoid the confusion.